profiling/service/service.go

```diff
@@ -0,0 +1,78 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package service
+
+import (
+	"context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+
+	"google.golang.org/grpc/internal/profiling"
+	ppb "google.golang.org/grpc/profiling/proto"
+	pspb "google.golang.org/grpc/profiling/proto/service"
+)
+
+type ProfilingConfig struct {
+	Enabled         bool
+	StreamStatsSize uint32
+	Server          *grpc.Server
+}
+
+func registerService(s *grpc.Server) {
+	grpclog.Infof("registering profiling service")
+	pspb.RegisterProfilingServer(s, &profilingServer{})
+}
+
+func Init(pc *ProfilingConfig) (err error) {
+	if err = profiling.InitStats(pc.StreamStatsSize); err != nil {
+		return
+	}
+
+	registerService(pc.Server)
+
+	// Do this last after everything has been initialised and allocated.
+	profiling.SetEnabled(pc.Enabled)
+
+	return
+}
+
+type profilingServer struct{}
+
+func (s *profilingServer) SetEnabled(ctx context.Context, req *pspb.SetEnabledRequest) (ser *pspb.SetEnabledResponse, err error) {
+	grpclog.Infof("processing SetEnabled (%v)", req.Enabled)
+	profiling.SetEnabled(req.Enabled)
+
+	ser = &pspb.SetEnabledResponse{Success: true}
```

Comment:
Do you have a valid case in mind where the `Success` field could end up with `false`? If not, consider using `google.protobuf.Empty`.

Reply:
Nope, this is a relic from weeks ago. I've removed the success field.

However, I propose not using `google.protobuf.Empty` as this may break forwards compatibility in the future in case we need to add new fields to the response. Switched to an empty `EnableResponse` message.